### PR TITLE
Add init function for language bundle and web context

### PR DESCRIPTION
**Pull request by jonathan-weibel-sonarsource**
[Edit 2020-04-24]
These changes imply the initialization of the sonar-ui-common library before doing anything else with it. There are 3 available functions that can be called separately or chained: 
```
import Initializer from 'sonar-ui-common/helpers/init';

Initializer
    .setLocale('en')      // language
    .setMessages({})      // translation strings
    .setUrlContext('');   // [required] web context
```
So, the web context can be setup before the l10n data, since you might need to use the request helper to get the language bundle (on SQ side). And then you can setup the locale and messages with the language bundle itself.

You can also see that the theme is not initialized by these functions. We could have done some hacking like we do for the tests to initialize the theme through this functions but that would break the React Context, because we would not use the context provider anymore and changes of the theme object would not be reflected anymore in the context consumer (we don't do that currently but we prefer avoid surprises in the future).

Not initializing the lib at all will throw an error as soon as something related to the web context is used and write console warning in dev mode when something from the language bundles is used. We added a default English language bundle with only one translation for when a request fail, so in dev mode you might get some warnings when first initializing the lib with only the web context to get your language bundle and this request fail, we think it's acceptable.

Remarks and questions for reviewers:
* We decided to update some of our dependencies too since we are gonna jump version
* The update of prettier changed nearly all the files, making the coverage break, we tried to add some quick win tests to improve it but it's not enough and we don't plan to spend more time adding coverage for these changes
* Most interesting commit to review and that should deserve your attention is **SC-1951 Add sonnar-ui-common initialization function**
* Do you think it would be better to have separate initialize functions for the web context and the language bundle ? [x] --> now implemented
 

**Checklist**

* [x] Functional validation
* [x] Documentation update
* [x] Update changelog

### CODE REVIEW

File: helpers/__tests__/init-test.ts
Comment (siegfried-ehret-sonarsource) [inline helpers/__tests__/init-test.ts]: You put `console.error` in `console.warn`. Is it on purpose ?

Comment (jonathan-weibel-sonarsource) [reply helpers/__tests__/init-test.ts]: OMG no! I switched to warn and forgot one replace. Thank you!

File: helpers/__tests__/init-test.ts
Comment (siegfried-ehret-sonarsource) [inline helpers/__tests__/init-test.ts]: Can you make explicit why we need this like ? Like describe [here](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-6.html#suppress-errors-in-ts-files-using--ts-ignore-comments).

File: helpers/__tests__/l10n-test.ts
Comment (siegfried-ehret-sonarsource) [inline helpers/__tests__/l10n-test.ts:20]: I removed this line and got no error when I ran eslint.

File: helpers/l10n.ts
Comment (siegfried-ehret-sonarsource) [inline helpers/l10n.ts]: Why not a `T.Dict<string>` ?

File: helpers/l10n.ts
Comment (siegfried-ehret-sonarsource) [inline helpers/l10n.ts]:
```suggestion
    // eslint-disable-next-line no-console
```

File: helpers/l10n.ts
Comment (siegfried-ehret-sonarsource) [inline helpers/l10n.ts]: Maybe it is just me but `L10n` brings noise to the variables and functions names. WDYT ?

File: config/jest/SetupSUC.ts
Comment (siegfried-ehret-sonarsource) [inline config/jest/SetupSUC.ts]: This is fine but I would like to know if you thought about an API like this:
```
initSonarUICommon({ urlContext: '' })
  .configureMessages({locale: 'en', messages: {}}); 
```
I don't think it would bring any value now, but depending on how we want to make sonar-ui-common configurable it would be nice to use something to kind of "namespace" the configuration object into the API. 
That would also help removing the `l10n` noise I mention in another comment.

File: components/charts/__tests__/ColorGradientLegend-test.tsx
Comment (siegfried-ehret-sonarsource) [inline components/charts/__tests__/ColorGradientLegend-test.tsx]:
```suggestion
it('should render properly', () => {
```

File: components/charts/__tests__/ZoomTimeLine-test.tsx
Comment (siegfried-ehret-sonarsource) [inline components/charts/__tests__/ZoomTimeLine-test.tsx:24]: Would it be possible to disallow the `jest/no-mocks-import` rules globally for test files ?

File: components/controls/__tests__/ReloadButton-test.tsx
Comment (siegfried-ehret-sonarsource) [inline components/controls/__tests__/ReloadButton-test.tsx]:
```suggestion
it('should render properly', () => {
```

File: components/ui/__tests__/GenericAvatar-test.tsx
Comment (siegfried-ehret-sonarsource) [inline components/ui/__tests__/GenericAvatar-test.tsx]:
```suggestion
it('should render properly', () => {
```

File: components/controls/__tests__/BackButton-test.tsx
Comment (siegfried-ehret-sonarsource) [inline components/controls/__tests__/BackButton-test.tsx]:
```suggestion
it('should render properly', () => {
```

File: components/controls/__tests__/HelpTooltip-test.tsx
Comment (siegfried-ehret-sonarsource) [inline components/controls/__tests__/HelpTooltip-test.tsx]:
```suggestion
it('should render properly', () => {
```

File: components/controls/__tests__/BackButton-test.tsx
Comment (siegfried-ehret-sonarsource) [inline components/controls/__tests__/BackButton-test.tsx]: The click would be better in another test, this one is only checking that it `should render properly`.

File: components/controls/__tests__/ReloadButton-test.tsx
Comment (siegfried-ehret-sonarsource) [inline components/controls/__tests__/ReloadButton-test.tsx:36]: Same here about the click to move to another test if possible.

File: components/controls/__tests__/SelectListListContainer-test.tsx
Comment (siegfried-ehret-sonarsource) [inline components/controls/__tests__/SelectListListContainer-test.tsx]: Comments can be removed, right ?

File: components/controls/__tests__/SelectListListContainer-test.tsx
Comment (siegfried-ehret-sonarsource) [inline components/controls/__tests__/SelectListListContainer-test.tsx]: Same here about the click to move to another test if possible.

File: components/controls/__tests__/ValidationModal-test.tsx
Comment (siegfried-ehret-sonarsource) [inline components/controls/__tests__/ValidationModal-test.tsx]: Same here about the submit to move to another test if possible.

File: components/intl/__tests__/DateTooltipFormatter-test.tsx
Comment (siegfried-ehret-sonarsource) [inline components/intl/__tests__/DateTooltipFormatter-test.tsx]:
```suggestion
  expect(shallowRender()).toMatchSnapshot();
```

**Review by siegfried-ehret-sonarsource — CHANGES_REQUESTED** [review]

**Comment by siegfried-ehret-sonarsource** [comment]
Great work on updating the deps and running prettier, thanks a lot! I am ok to not care about coverage now.

I added a few suggestions and a few questions. It is in "request changes" because I will run SonarCloud with this sonar-ui-common tomorrow.

File: helpers/__tests__/init-test.ts
Comment (gregaubert) [reply helpers/__tests__/init-test.ts]: Good idea! We want to have everything initialized to undefined in this test, it's not possible if we respect the types

File: README.md
Comment (jeremy-davis-sonarsource) [inline README.md]: Typo: 
> To use the library it should first ~~with~~ be initialized by calling the following:

File: helpers/__tests__/l10n-test.ts
Comment (gregaubert) [reply helpers/__tests__/l10n-test.ts:20]: This is definitely needed, there is a lot of translation keys that don't respect the camelcase rule in this file, we don't want to have them reported on SQ.

This rule raise warnings, not errors, so you won't see them when running `yarn lint`

File: config/jest/SetupSUC.ts
Comment (siegfried-ehret-sonarsource) [reply config/jest/SetupSUC.ts]: Thinking again about this, we could also inject the theme in some API.

File: components/charts/__tests__/ZoomTimeLine-test.tsx
Comment (gregaubert) [reply components/charts/__tests__/ZoomTimeLine-test.tsx:24]: I think this rule apply only to test files, it's from jest plugin. I think the issue here is more that this theme file should maybe not be in a mock folder, because it doesn't mock anything that exists in SUC. So it's not a mock, it's just a default theme used for tests, but I don't know where we should put it...

Comment (siegfried-ehret-sonarsource) [reply components/charts/__tests__/ZoomTimeLine-test.tsx:24]: Ok don't worry about it ! Thanks.

File: config/jest/SetupSUC.ts
Comment (gregaubert) [reply config/jest/SetupSUC.ts]: Yeah there is probably way to improve the API, we tried to keep it simple there. But maybe it's better to have separate initialization function. But I feel like the API you suggested might be overly complex for what we need, it would mean return a class instance that has init functions that would set a global variable ?

Regarding the theme:
> You can also see that the theme is not initialized by this function, we could have do some hacking like we do for the tests to initialize the theme through this function but that would break the React Context, because we would not use the context provider anymore and changes of the theme object would not be reflected anymore in the context consumer (we don't do that currently but we prefer avoid surprises for the future).

Comment (gregaubert) [reply config/jest/SetupSUC.ts]: I'd be happy to hear about the API from @jeremy-davis-sonarsource and @philippe-perrin-sonarsource too before we change something

File: helpers/l10n.ts
Comment (gregaubert) [reply helpers/l10n.ts]: True, I'll drop the L10n part!

File: config/jest/SetupSUC.ts
Comment (siegfried-ehret-sonarsource) [reply config/jest/SetupSUC.ts]: Simple is good ! :)
Yeah it is more complex, it would bring a class with methods that return the instance etc...
I'll just repeat one thing from my previous message: "I don't think it would bring any value now", there is no need to change anything but we can think about it during a Web coffee break !

File: helpers/__tests__/urls-test.ts
Comment (jeremy-davis-sonarsource) [inline helpers/__tests__/urls-test.ts:48]: NP: Add a test to check that `getHostUrl` throws as expected?

**Review by jeremy-davis-sonarsource — APPROVED** [review]
Wow, great job! This is going to be a good excuse to fix our bootstrapping code too...
Just a few NPs

Did a quick clutch to make it work with SQ, no issues detected

File: config/jest/SetupSUC.ts
Comment (jeremy-davis-sonarsource) [reply config/jest/SetupSUC.ts]: I like the dissociation of the urlContext from the locale & messages. Maybe we could expose a plain object with references to individual functions ?
The setters could return the object, for chaining:

```
import initializer from 's-u-c/helpers/init'

initializer.setUrlContext('/');
initializer
  .setMessages({...})
  .setLocale('en');

```

Comment (jonathan-weibel-sonarsource) [reply config/jest/SetupSUC.ts]: I like this design too, I'll give a try!

**Comment by jonathan-weibel-sonarsource** [comment]
@gregaubert @jeremy-davis-sonarsource @siegfried-ehret-sonarsource I updated the initialization, using a small class with static methods. Could you have a look? Worked with SonarCloud.

**Comment by jonathan-weibel-sonarsource** [comment]
> @gregaubert @jeremy-davis-sonarsource @siegfried-ehret-sonarsource I updated the initialization, using a small class with static methods. Could you have a look? Worked with SonarCloud.

The sample use in SC is in the third commit of this PR: https://github.com/SonarSource/sonarcloud-core/pull/861

**Comment by siegfried-ehret-sonarsource** [comment]
Hey @jonathan-weibel-sonarsource maybe you don't need a class, and if you want one I'd do something like [this](https://github.com/SonarSource/sonar-ui-common/commit/fe1f05a6232642eef90f13f64b70000a2fc56812) (I haven't touched the tests and I don't know if it works ^^)

**Comment by siegfried-ehret-sonarsource** [comment]
And it is also very fine how it was with only a function and a parameter object with everything, this is really an idea we can discuss and implement later.

**Comment by jeremy-davis-sonarsource** [comment]
> Hey @jonathan-weibel-sonarsource maybe you don't need a class, and if you want one I'd do something like [this](https://github.com/SonarSource/sonar-ui-common/commit/fe1f05a6232642eef90f13f64b70000a2fc56812) (I haven't touched the tests and I don't know if it works ^^)

Yeah I was thinking no class as well just a plain:
```
const init = {
   setUrlContext = (url) => {urlContext = url ; return init; }
}
```

**Comment by gregaubert** [comment]
Yeah I don't think we need a class, I'll simplify a bit :) Thanks for the feedback, I think it's better indeed with separate initializer!

**Comment by gregaubert** [comment]
I'm gonna do a RC release of this so we can have the full QA running on SC side, but won't merge yet

**Comment by gregaubert** [comment]
@philippe-perrin-sonarsource @jeremy-davis-sonarsource @siegfried-ehret-sonarsource I plan to do the actual release and merge this PR tomorrow (morning). So please raise the hand if you think there are some more changes needed or if you need more time to have a last look! :eyes:

**Comment by philippe-perrin-sonarsource** [comment]
@gregaubert I need more time (was off last friday and busy with review/meetings today). Go ahead if you're fine with the review you've got so far. Otherwise, that's my top priority for tomorrow morning.

**Comment by gregaubert** [comment]
> @gregaubert I need more time (was off last friday and busy with review/meetings today). Go ahead if you're fine with the review you've got so far. Otherwise, that's my top priority for tomorrow morning.

Great :+1: , I'll wait for your review tomorrow morning before merging (or not) then!

**Review by philippe-perrin-sonarsource — APPROVED** [review]
LGTM. Didn't tested it though. I think all the relevant comments has been made.

Good job guys.

**Pull request merged**
